### PR TITLE
feat: replace `opis/closure` with `laravel/serializable-closure`

### PR DESCRIPTION
**Pull request by owenvoke**
Looks like this would require a dependency bump to `^7.4 || ^8.0` for Box, as unserialisation requires PHP 7.4 ([`SerializableClosure.php`](https://github.com/laravel/serializable-closure/blob/master/src/SerializableClosure.php#L28))...

I noticed that one of the other dependencies ([`amphp/parallel-functions`](https://github.com/amphp/parallel-functions)) also uses Opis Closure. ~~I'll also open a PR to update that and see if they're happy~~ (I've opened https://github.com/amphp/parallel-functions/pull/29). 👍🏻 However, I'm not entirely sure that it would be merged as the Laravel one only supports PHP `^7.3 || ^8.0` and `parallel-functions` is currently `>=7`

Related issue: https://github.com/box-project/box/issues/571

### CODE REVIEW

**Comment by owenvoke** [comment]
Lol. Off to a good start, I'll investigate the issues with this, clearly not as easy as a direct switch... 😬

**Comment by owenvoke** [comment]
I think some of the failures may be related to PHPScoper (see the PHP 8.1 test suites). 🤔 There's also a missing polyfill for `str_contains()` in [this test run](https://github.com/box-project/box/runs/4853178677?check_suite_focus=true), which I'm not sure why as I've included the polyfill in Composer. 🤔

The PHP 7.4 E2E errors are because of the `str_contains()` missing (which was added in PHP 8.0), and the PHP 8.0 E2E tests are failing because of https://github.com/box-project/box/runs/4853179758?check_suite_focus=true#step:10:131 which I'm not sure how to debug...

**Comment by theofidry** [comment]
@owenvoke I've fixed the polyfills issues. However I don't recommend to invest more time in this PR until the parallel-functions one is sorted as I'm likely gonna change more things which would result in more conflicts

**Comment by owenvoke** [comment]
https://github.com/amphp/parallel-functions/pull/29 has been merged now. 🥳 I'll wait for a release of that, and then rebase / update this PR. 👍🏻

**Comment by owenvoke** [comment]
This has been rebased and is working, just waiting on the tagged release of `amphp/parallel-functions` 👍🏻 Looks like the E2E tests are also failing on PHP 8.1 due to `phpspec/prophecy` not supporting PHP 8.1 yet.

**Comment by theofidry** [comment]
>Looks like the E2E tests are also failing on PHP 8.1 due to phpspec/prophecy-phpunit not supporting v2.x of phpspec/prophecy yet.

Yes this is fine. I already enabled the ignore platform req php+ flag but didn't manage to find a nice way to propagate it to the makefile as well... So some tests are bound to fail this way it is ok

File: tests/RequirementChecker/RequirementsDumperTest.php [hunk 1]
```diff
@@ -113,7 +113,7 @@ public function provideJsonAndLockContents(): Generator
                     [
                         'name' => 'acme/foo',
                         'require' => [
-                            'php' => '^7.3',
+                            'php' => '^7.4',
```
Comment (theofidry) [inline tests/RequirementChecker/RequirementsDumperTest.php:116]: are those changes really necessary?

File: src/PhpScoper/SimpleScoper.php [hunk 2]
```diff
@@ -125,7 +125,7 @@ private function getScoper(): PhpScoper
     /**
      * @param callable[] $patcher
      *
-     * @retunr SerializableClosure[]
+     * @return SerializableClosure[]
```
Comment (theofidry) [inline src/PhpScoper/SimpleScoper.php:128]: nice catch

File: composer.json [hunk 3]
```diff
@@ -44,15 +44,15 @@
         "php": "^7.4 || ^8.0",
         "ext-phar": "*",
         "composer-plugin-api": "^2.2",
-        "amphp/parallel-functions": "^1.0",
+        "amphp/parallel-functions": "dev-master",
```
Comment (theofidry) [inline composer.json:47]: Although this good for running the tests I would prefer to wait for the release before merging the PR

File: tests/RequirementChecker/RequirementsDumperTest.php (hunk 1 above)
Comment (owenvoke) [reply tests/RequirementChecker/RequirementsDumperTest.php:116]: No, I guess not. 👍🏻 Just was thinking that test was for the lowest PHP version that's supported.

File: composer.json (hunk 3 above)
Comment (owenvoke) [reply composer.json]: Updated to `^1.1` now 🥳

File: tests/RequirementChecker/RequirementsDumperTest.php (hunk 1 above)
Comment (owenvoke) [reply tests/RequirementChecker/RequirementsDumperTest.php:116]: Do you want me to revert the changes to this file?

Comment (theofidry) [reply tests/RequirementChecker/RequirementsDumperTest.php:116]: No it's fine. I was just confused by it initally

**Comment by theofidry** [comment]
Thanks a lot @owenvoke!

**Review by Alexanderndi — COMMENTED** [review]
how do i get the serializableClosure installed

**Comment by theofidry** [comment]
This is purely internal to Box @Alexanderndi

**Pull request merged**